proposals/readonly-methods.md

```diff
@@ -0,0 +1,65 @@
+# Readonly Methods
+
+* [x] Proposed
+* [ ] Prototype: Not Started
+* [ ] Implementation: Not Started
+* [ ] Specification: Not Started
+
+## Summary
+[summary]: #summary
+
+Provide a way to specify individual methods on a struct do not modify state, in the same way that `readonly struct` specifies no instance method modifies state.
+
+It is worth noting that `readonly method` != `pure method`. A `pure` method guarantees no state will be modified. A `readonly` method only guarantees that instance state will not be modified.
+
+All instance methods on a `readonly struct` could be considered implicitly `readonly methods`. Explicit `readonly methods` declared on non-readonly structs would behave in the same manner. For example, they would still create hidden copies if you called an instance method (on the current instance or on a field of the instance) which was itself not-readonly.
+
+## Motivation
+[motivation]: #motivation
+
+Today, users have the ability to create `readonly struct` types which the compiler enforces that all fields are readonly (and by extension, that no methods modify the state). However, there are some scenarios where you have an existing API that exposes public fields, or needs to modify some small amount of internal state in select scenarios. Under these circumstances, you cannot mark the type as `readonly` (it would be a breaking change).
+
+This normally doesn't have much impact, except in the case of `in` parameters. With `in` parameters for non-readonly structs, the compiler will make a copy of the parameter for each instance method call, since it cannot guarantee that the call does not modify internal state. This can lead to a multitude of copies and worse overall performance than if you had just passed the struct directly by value. For an example, see this code on [sharplab](https://sharplab.io/#v2:CYLg1APgAgDABFAjAbgLACgNQMxwM4AuATgK4DGBcAagKYUD2RATBgN4ZycK4BmANvQCGlAB5p0XbnH5DKAT3GSOXHNIHC4AGRoA7AOYEAFgGUAjiUFEawZZ3YTJXPTQK3H9x54QB2OAAoROAAqOBEASjgwNy8YvzlguDkwxS8AXzd09EysXCgmOABhOA8VXnVKAFk/AEsdajoCRnyAN0E+EhoIks8oX1b2mgA6bX0jMwsrYEi4fo7h3QMTc0trFM5M1KA==)
+
+## Design
+[design]: #design
+
+I propose allowing a user to specify that a method itself is `readonly` and does not modify the state of the struct (with all the appropriate verification done by the compiler, of course). For example:
+
+```C#
+public struct Vector2
+{
+    public float x;
+    public float y;
+
+    public float GetLength() readonly
+    {
+        return MathF.Sqrt(LengthSquared);
```

Comment:
I think it would help to give a negative example that demonstrates exactly what this does. Essentially marking a method `readonly` and showing there is a compilation error when attempting to write to a mutable field.

Reply:
Fixed.